app/models/katello/glue/provider.rb

```diff
@@ -123,63 +123,23 @@ def import_logger
-          product_content_attrs = attrs.delete(:productContent) || []
-          product = ::Katello::Product.find(id_pair.first)
-          Glue::Candlepin::Product.import_product_content(product, product_content_attrs)
-        end
-      end
-
-      # TODO: break up method
-      def import_products_from_cp # rubocop:disable MethodLength
-        db_and_cp_ids = self.organization.providers.redhat.first.products.pluck(:id, :cp_id)
-
-        # this method only imports products which are not already in Katello
-        # no need for a full import on them - just update the product content
-        update_product_content_from_cp(db_and_cp_ids)
-
-        product_in_katello_ids = db_and_cp_ids.map(&:second)
-        products_in_candlepin_ids = []
-
-        marketing_to_engineering_product_ids_mapping.each do |marketing_product_id, engineering_product_ids|
-          engineering_product_ids = engineering_product_ids.uniq
-          products_in_candlepin_ids << marketing_product_id
-          products_in_candlepin_ids.concat(engineering_product_ids)
-          added_eng_products = (engineering_product_ids - product_in_katello_ids).map do |id|
-            Resources::Candlepin::Product.get(self.organization.label, id)[0]
-          end
-          adjusted_eng_products = []
-          added_eng_products.each do |product_attrs|
-            begin
-              Glue::Candlepin::Product.import_from_cp(product_attrs) do |p|
-                p.provider = self
-                p.organization_id = self.organization.id
-              end
-              adjusted_eng_products << product_attrs
-              import_logger.info "import of product '#{product_attrs["name"]}' from Candlepin OK"
-            rescue Errors::SecurityViolation => e
-              # Do not add non-accessible products
-              logger.info "import of product '#{product_attrs["name"]}' from Candlepin failed"
-              import_logger.info e
-            end
-          end
-
-          product_in_katello_ids.concat(adjusted_eng_products.map { |p| p["id"] })
-
-          marketing_product = Katello::Product.find_by_cp_id(marketing_product_id)
-          marketing_product.destroy if marketing_product && marketing_product.redhat?
-        end
+      def import_products_from_cp
+        cp_products = ::Katello::Resources::Candlepin::Product.all(organization.label, [:id, :name, :multiplier, :productContent])
+        cp_products = cp_products.select { |prod| Glue::Candlepin::Product.engineering_product_id?(prod['id']) }
 
-        product_to_remove_ids = (product_in_katello_ids - products_in_candlepin_ids).uniq
-        product_to_remove_ids.each do |cp_id|
-          product = Product.find_by_cp_id(cp_id, self.organization)
-          Rails.logger.warn "Orphaned Product id #{product.id} found while refreshing/importing manifest."
-        end
+        cp_products.each { |product| import_product(product) }
```

Comment:
Could this be made concurrent in the action? Would it improve performance?

Reply:
i'm not sure if i like that direction here, as it would create up to ~100s of subtasks, but i will give it a shot and see

Reply:
actually, thats really difficult to do here, because we can't plan each step beforehand because we don't know what products exist until after the manifest has been imported or refreshed

Reply:
Ok! It'd most likely would be slower just creating all those records, as you said.